server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -465,6 +466,18 @@ public Decision canRemain(IndexMetadata indexMetadata, ShardRouting shardRouting
         );
     }
 
+    private static boolean ignoreDiskWatermarksForShard(
+        RoutingAllocation allocation,
+        IndexMetadata indexMetadata,
+        ShardRouting shardRouting
+    ) {
+        return indexMetadata.ignoreDiskWatermarks() || ignoreDiskWatermarksForShard(allocation, shardRouting);
+    }
```

Comment:
In case when `ignoreDiskWatermarks()==false` this would check supplied indexMetadata, go to ` ignoreDiskWatermarksForShard` to check the role and then load `indexMetadata` again to check `ignoreDiskWatermarks` on it.  I do not think it is necessary to load the indexMetadata and check it again.

This could be:
```
    private static boolean ignoreDiskWatermarksForShard(RoutingAllocation allocation, IndexMetadata indexMetadata) {
        return shardRouting.role() == Role.INDEX_ONLY || indexMetadata.ignoreDiskWatermarks();
    }

    private static boolean ignoreDiskWatermarksForShard(RoutingAllocation allocation, ShardRouting shardRouting) {
        return ignoreDiskWatermarksForShard(allocation, allocation.metadata().index(shardRouting.index()));
    }
``

This way the first one would only check already present IndexMetadata once regardless `ignoreDiskWatermarks` the return value and could avoid extra `allocation.metadata().index(shardRouting.index())` lookup.

Reply:
Is 3e957cf1f5a072a7f96ea1653286b999240ca9c0 better?

Reply:
I'm ok with how it is now, although I'd kinda prefer to skip the lookup in `allocation.metadata().index(shardRouting.index())` if `shardRouting.role() == Role.INDEX_ONLY`. These lookups do add up at high index counts.

Reply:
Yeah, that's how I did it at first but reviews seem to push in a different direction :)